### ISSUE TITLE
Pilot configures proxy with wrong routes/listeners

### ISSUE THREAD

**Issue by jasonwoodman-ascend** [istio/istio#20733]
**Bug description**
Pilot adds incorrect listeners/routes to envoy proxies, resulting in 503's.  This is the output of proxy-config routes on 2 identical pods created from the same deployment:

This is the proxy that received the correct config:
```
$ istioctl pc routes pod1.our-namespace
NOTE: This output only contains routes loaded via RDS.
NAME                                                                                                    VIRTUAL HOSTS
8080                                                                                                    4
9091                                                                                                    2
istio-telemetry.istio-system.svc.cluster.local:42422                                                    1
15004                                                                                                   2
15014                                                                                                   2
inbound|8080|http|correct-service.our-namespace.svc.cluster.local     1
inbound|8080|http|correct-service.our-namespace.svc.cluster.local     1
inbound|8080|http|correct-service.our-namespace.svc.cluster.local     1
inbound|8080|http|correct-service.our-namespace.svc.cluster.local     1
```
Here is the output from an incorrectly configured proxy on the other pod from the same deployment as above:
```
$ istioctl pc routes pod2.our-namespace
NOTE: This output only contains routes loaded via RDS.
NAME                                                           VIRTUAL HOSTS
8080                                                           4
istio-telemetry.istio-system.svc.cluster.local:42422           1
9091                                                           2
15004                                                          2
15014                                                          2
inbound|8800|http|service-from-another-namespace.wrong-namespace.svc.cluster.local     1
inbound|8800|http|service-from-another-namespace.wrong-namespace.svc.cluster.local     1
inbound|8800|http|service-from-another-namespace.wrong-namespace.svc.cluster.local     1
inbound|8800|http|service-from-another-namespace.wrong-namespace.svc.cluster.local     1
```
Also notice the port number is not right either.  Restarting the second pod will result in a proxy with the same config as the first, which is correct.

**Expected behavior**
Pilot configures proxy with correct routes/listeners for the service.  

**Steps to reproduce the bug**
It seems to only happen on a deployment when new pods are created and not every time.  Restarting the pods seems to yield correct listeners every time.

**Version (include the output of `istioctl version --remote` and `kubectl version` and `helm version` if you used Helm)**
```
$ istioctl version --remote
client version: 1.4.3
citadel version: 1.4.3
galley version: 1.4.3
galley version: 1.4.3
istio-ig-custom version: 
istio-ig-custom version: 
istio-ig-custom2 version: 
istio-ig-custom2 version: 
istio-ig-custom3 version: 
istio-ig-custom3 version: 
pilot version: 1.4.3
pilot version: 1.4.3
sidecar-injector version: 1.4.3
sidecar-injector version: 1.4.3
telemetry version: 1.4.3
telemetry version: 1.4.3
data plane version: 1.4.0 (1 proxies), 1.4.3 (229 proxies)

$ kubectl version
Client Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.3", GitCommit:"b3cbbae08ec52a7fc73d334838e18d17e8512749", GitTreeState:"clean", BuildDate:"2019-11-14T04:24:34Z", GoVersion:"go1.12.13", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.5", GitCommit:"20c265fef0741dd71a66480e35bd69f18351daea", GitTreeState:"clean", BuildDate:"2019-10-15T19:07:57Z", GoVersion:"go1.12.10", Compiler:"gc", Platform:"linux/amd64"}
```

**How was Istio installed?**
Helm 2 with Tiller

**Environment where bug was observed (cloud vendor, OS, etc)**
On premise, centos

**Comment by howardjohn** [comment]
Please give the full output of config dump https://github.com/istio/istio/wiki/Troubleshooting-Istio and what inputs (Service, virtual service, etc) produce this

**Comment by pmoncadaisla** [comment]
We are seeing the same or simular issue randomly in our pod **after upgrading from Istio 1.3.3 to 1.4.3**


**Steps to reproduce the bug**
It seems to only happen on a deployment when new pods are created and not every time. Restarting the pods seems to yield correct listeners every time.


**Version (include the output of istioctl version --remote and kubectl version and helm version if you used Helm)**
```
$ istioctl version --remote

client version: 1.4.3
citadel version: 1.4.3
egressgateway version: 1.4.3
egressgateway version: 1.4.3
egressgateway version: 1.4.3
egressgateway version: 1.4.3
egressgateway version: 1.4.3
galley version: 1.4.3
ilbgateway version: 1.4.3
ilbgateway version: 1.4.3
ingressgateway version: 1.4.3
ingressgateway version: 1.4.3
ingressgateway version: 1.4.3
ingressgateway version: 1.4.3
ingressgateway version: 1.4.3
nodeagent version:
[... nodeagent version ...]
nodeagent version:
pilot version: 1.4.3
pilot version: 1.4.3
policy version: 1.4.3
sidecar-injector version: 1.4.3
telemetry version: 1.4.3
telemetry version: 1.4.3
data plane version: 1.4.3 (838 proxies)
```

```
$ kubectl version

Client Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.2", GitCommit:"59603c6e503c87169aea6106f57b9f242f64df89", GitTreeState:"clean", BuildDate:"2020-01-18T23:30:10Z", GoVersion:"go1.13.5", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"14+", GitVersion:"v1.14.7-gke.23", GitCommit:"81c87c699557fed991e292cd328b2129c2f242a2", GitTreeState:"clean", BuildDate:"2019-11-07T19:23:23Z", GoVersion:"go1.12.11b4", Compiler:"gc", Platform:"linux/amd64"}
```

**How was Istio installed?**
Helm 2 with Tiller

**Environment where bug was observed (cloud vendor, OS, etc)**

- Google Cloud
- GKE 
```
$ gcloud container clusters list
NAME           LOCATION      MASTER_VERSION  MASTER_IP     MACHINE_TYPE   NODE_VERSION   NUM_NODES  STATUS
a-cluster      europe-west1  1.14.7-gke.23   x.x.x.x.x     n1-standard-4  1.14.7-gke.23  108        RUNNING
```
### Additional information about this issue:

These are the routes in a normal pod (edited the names, I can provide the output wihtout edits via a private channel):
```
$ istioctl pc routes a-deployment-794f446794-zzwvm.correct-namespace
NOTE: This output only contains routes loaded via RDS.
NAME                                                                                                       VIRTUAL HOSTS
80                                                                                                         33
96                                                                                                         2
5556                                                                                                       2
8060                                                                                                       2
8080                                                                                                       10
8081                                                                                                       3
9090                                                                                                       3
9091                                                                                                       30
9901                                                                                                       2
10901                                                                                                      2
15004                                                                                                      3
15010                                                                                                      2
15014                                                                                                      7
15030                                                                                                      2
15031                                                                                                      2
jaeger-istio-collector.istio-system.svc.cluster.local:9411                                                 1
jaeger-istio-collector.istio-system.svc.cluster.local:14267                                                1
inbound|9091|http-prom-a-deployment-correct-namespace|a-deployment.correct-namespace.svc.cluster.local     1
jaeger-istio-collector.istio-system.svc.cluster.local:14268                                                1
istio-telemetry.istio-system.svc.cluster.local:42422                                                       1
inbound|9091|http-prom-a-deployment-correct-namespace|a-deployment.correct-namespace.svc.cluster.local     1
inbound|80|http-a-deployment-correct-namespace|a-deployment.correct-namespace.svc.cluster.local            1
inbound|80|http-a-deployment-correct-namespace|a-deployment.correct-namespace.svc.cluster.local            1
20001                                                                                                      2
                                                                                                           1
```

This are the routes from a faulty pod 
```
istioctl pc routes a-deployment-794f446794-clvbg.correct-namespace
NOTE: This output only contains routes loaded via RDS.
NAME                                                                                VIRTUAL HOSTS
80                                                                                  33
96                                                                                  2
5556                                                                                2
8060                                                                                2
istio-telemetry.istio-system.svc.cluster.local:42422                                1
jaeger-istio-collector.istio-system.svc.cluster.local:14268                         1
inbound|80|http-other-deployment-wrong-namespace|other-deployment.wrong-namespace.svc.cluster.local     1
jaeger-istio-collector.istio-system.svc.cluster.local:9411                          1
jaeger-istio-collector.istio-system.svc.cluster.local:14267                         1
inbound|80|http-other-deployment-wrong-namespace|other-deployment.wrong-namespace.svc.cluster.local     1
8080                                                                                10
8081                                                                                3
9090                                                                                3
9091                                                                                30
9901                                                                                2
10901                                                                               2
15004                                                                               3
15010                                                                               2
15014                                                                               7
15030                                                                               2
15031                                                                               2
20001                                                                               2
                                                                                    1
```

When executing an http request to a POD with the correct configuration we can see the header`x-envoy-decorator-operation` set to the correct POD name:
```
curl -I http://a-deployment-794f446794-zzwvm.correct-namespace:8080
< server: istio-envoy
< x-envoy-decorator-operation: a-deployment.correct-namespace:80/*

```

If we execute the same request to a POD with a wrong configuration we see:
```
curl -I http://a-deployment-794f446794-clvbg.correct-namespace:8080
< server: istio-envoy
< x-envoy-decorator-operation: other-deployment.wrong-namespace:80/*

```

where you can notice the `x-envoy-decorator-operation` header indicates `other-deployment.wrong-namespace:80` but we expected to see `a-deployment-794f446794-clvbg.correct-namespace:80`

Additionally, we have detected that the content response is correct, it is provided by `a-deployment-794f446794-clvbg.correct-namespace` but Policies are wrong (in our case, JWT Policy).

### Kubernetes and Istio resources for `a-deployment` in `correct-namespace`:

**VirtualService**:
```
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  creationTimestamp: "2019-08-07T06:15:32Z"
  generation: 3
  labels:
    app: a-deployment
    chart: base-v1.2.17
    fullapp: correct-namespace
    heritage: Tiller
    release: a-deployment-correct-namespace
    version: v1.2.17
  name: a-deployment
  namespace: correct-namespace
  resourceVersion: "281486396"
  selfLink: /apis/networking.istio.io/v1alpha3/namespaces/correct-namespace/virtualservices/a-deployment
  uid: c3360aef-b8da-11e9-b4e4-42010ac50709
spec:
  gateways:
  - default/ingressgateway
  hosts:
  - correct-namespace.dev-01.my.domain.com
  - other.dev.my.domain.com
  http:
  - corsPolicy:
      allowCredentials: false
      allowHeaders:
      - origin
      - x-requested-with
      - accept
      - content-type
      - x-application-id
      - x-correlation-id
      - x-request-id
      - authorization
      - x-authorization
      - X-Customer-ID
      - X-SFID
      - X-MSISDN
      - x-front-client
      - scoring-token
      - x-target
      - cache-control
      allowOrigin:
      - '*'
      maxAge: 600s
    match:
    - uri:
        prefix: /v1/a-deployment/
    rewrite:
      uri: /
    route:
    - destination:
        host: a-deployment.correct-namespace.svc.cluster.local
        port:
          number: 80
```

**Service**:

```
piVersion: v1
kind: Service
metadata:
  annotations:
    prometheus.io/path: /metrics
    prometheus.io/port: "9091"
    prometheus.io/probe: "true"
  creationTimestamp: "2019-08-07T06:15:32Z"
  labels:
    app: a-deployment
    chart: base-v1.2.17
    fullapp: correct-namespace
    heritage: Tiller
    release: a-deployment-correct-namespace
    version: v1.2.17
  name: a-deployment
  namespace: correct-namespace
  resourceVersion: "435659168"
  selfLink: /api/v1/namespaces/correct-namespace/services/a-deployment
  uid: c3314118-b8da-11e9-b4e4-42010ac50709
spec:
  clusterIP: 10.197.12.204
  ports:
  - name: http-a-deployment-correct-namespace
    port: 80
    protocol: TCP
    targetPort: 8080
  - name: http-prom-a-deployment-correct-namespace
    port: 9091
    protocol: TCP
    targetPort: 9091
  selector:
    app: a-deployment
    release: a-deployment-correct-namespace
  sessionAffinity: None
  type: ClusterIP

```

**Policy**

```
apiVersion: authentication.istio.io/v1alpha1
kind: Policy
metadata:
  creationTimestamp: "2019-10-21T14:11:59Z"
  generation: 1
  labels:
    app: a-deployment
    chart: base-v1.2.17
    fullapp: correct-namespace
    heritage: Tiller
    release: a-deployment-correct-namespace
    tier: backend
    version: v1.2.17
  name: a-deployment-authn
  namespace: correct-namespace
  resourceVersion: "281486392"
  selfLink: /apis/authentication.istio.io/v1alpha1/namespaces/correct-namespace/policies/a-deployment-authn
  uid: bf1eaa37-f40c-11e9-8d86-42010ac50709
spec:
  origins:
  - jwt:
      issuer: issuer1
      jwksUri: https://auth-provider-jwks.example.com/.well-known/jwks.json
      trigger_rules:
      - excluded_paths:
        - exact: /health
        - prefix: /docs
        - exact: /health
        - exact: /metrics
        included_paths:
        - prefix: /
  - jwt:
      issuer: issuer2
      jwksUri: https://auth-provider-jwks.example.com/.well-known/jwks.json
      trigger_rules:
      - excluded_paths:
        - exact: /health
        - prefix: /docs
        - exact: /health
        - exact: /metrics
        included_paths:
        - prefix: /
  principalBinding: USE_ORIGIN
  targets:
  - name: a-deployment
```

### Kubernetes and Istio resources for `other-deployment` in `wrong-namespace`:

**VirtualService**
```
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  creationTimestamp: "2019-08-14T10:09:51Z"
  generation: 3
  labels:
    app: other-deployment
    chart: base-v1.2.17
    fullapp: wrong-namespace
    heritage: Tiller
    release: other-deployment-wrong-namespace
    version: v1.2.17
  name: other-deployment
  namespace: wrong-namespace
  resourceVersion: "294862447"
  selfLink: /apis/networking.istio.io/v1alpha3/namespaces/wrong-namespace/virtualservices/other-deployment
  uid: a7b23296-be7b-11e9-bf65-42010ac5070b
spec:
  gateways:
  - default/ingressgateway
  hosts:
  - wrong-namespace.dev.my.domain.com
  - other-url.dev.my.domain.com
  http:
  - corsPolicy:
      allowCredentials: false
      allowHeaders:
      - origin
      - x-requested-with
      - accept
      - content-type
      - x-requestId
      - x-requestSessionId
      - authorization
      - cache-control
      allowOrigin:
      - '*'
      maxAge: 600s
    match:
    - uri:
        prefix: /other-deployment/
    rewrite:
      uri: /
    route:
    - destination:
        host: other-deployment.wrong-namespace.svc.cluster.local
        port:
          number: 80
```

**Service**
```
apiVersion: v1
kind: Service
metadata:
  annotations:
  creationTimestamp: "2019-08-14T10:09:51Z"
  labels:
    app: other-deployment
    chart: base-v1.2.17
    fullapp: wrong-namespace
    heritage: Tiller
    release: other-deployment-wrong-namespace
    version: v1.2.17
  name: other-deployment
  namespace: wrong-namespace
  resourceVersion: "435658730"
  selfLink: /api/v1/namespaces/wrong-namespace/services/other-deployment
  uid: a7afea06-be7b-11e9-bf65-42010ac5070b
spec:
  clusterIP: 10.197.10.227
  ports:
  - name: http-other-deployment-wrong-namespace
    port: 80
    protocol: TCP
    targetPort: 8080
  selector:
    app: other-deployment
    release: other-deployment-wrong-namespace
  sessionAffinity: None
  type: ClusterIP
```

**Policy**
```
apiVersion: authentication.istio.io/v1alpha1
kind: Policy
metadata:
  creationTimestamp: "2019-09-11T10:26:43Z"
  generation: 1
  labels:
    app: other-deployment
    chart: base-v1.2.17
    fullapp: wrong-namespace
    heritage: Tiller
    release: other-deployment-wrong-namespace
    tier: backend
    version: v1.2.17
  name: other-deployment-authn
  namespace: wrong-namespace
  resourceVersion: "294862440"
  selfLink: /apis/authentication.istio.io/v1alpha1/namespaces/wrong-namespace/policies/other-deployment-authn
  uid: a64ba2ce-d47e-11e9-bf65-42010ac5070b
spec:
  origins:
  - jwt:
      issuer: issuer3
      jwksUri: https://auth-provider-jwks.example.com/.well-known/jwks.json
      trigger_rules:
      - excluded_paths:
        - exact: /monitoring/health
        - prefix: /doc/
  principalBinding: USE_ORIGIN
  targets:
  - name: other-deployment
```

As for the config dump file, it is attached.

[istio-config-dump-edited.json.gz](https://github.com/istio/istio/files/4160801/istio-config-dump-edited.json.gz)

I hope this information is useful.

Thanks!

**Comment by howardjohn** [comment]
This may just be some recency bias, but this feels a lot like https://github.com/istio/istio/issues/20676. tl;dr is a pod IP gets re-used, causing pilot to generate config for the old pod instead of the new pod. So presumably, if this was the case, there was a pod with some ip 1.2.3.4 in service-from-another-namespace.wrong-namespace, then later your new pod in service correct-service.our-namespace comes up with pod ip 1.2.3.4 and gets the wrong config. I think this can only happen with pre-emptible nodes though, are you using those?

The tell-tale sign of this is if you see this pattern in the logs:
```
Handling event update for pod pod1-sha123 in namespace foo -> 1.2.3.4
Handling event update for pod pod1-sha123 in namespace foo -> 1.2.3.5 # same pod name, new IP
Handling event update for pod new-pod-sha345 in namespace foo -> 1.2.3.4 # re-use the old pod
```

**Comment by jasonwoodman-ascend** [comment]
@howardjohn In our case the nodes are long-living, they are on-prem and have been running for nearly a year.

**Comment by pmoncadaisla** [comment]
Hi @howardjohn , we are indeed using preemtible nodes (and not preemtible).

I've search for "Handling event update for pod" in Stackdriver logs (whole cluster) and I couldn't find a single match. (Also tested substrings)

**Comment by howardjohn** [comment]
@pmoncadaisla if you are using pre-emptible nodes it seems likely the root cause. Thats weird it doesn't match, that log happens any time a pod is changed at all, do you not have `info` level logging enabled?

**Comment by pmoncadaisla** [comment]
@howardjohn I can see this in the `istio-proxy` container of any POD: `--proxyLogLevel="warning"` and `--log_output_level="default:info"`

Pre-emptible notes seems to me the root cause, but, why is this happening after the upgrade from `1.3.3` to `1.4.3` and not before? We've seem this exact behaviour in 2 different Kubernetes clusters, upgraded from and to the same Istio version.

**Comment by howardjohn** [comment]
My understand was that this code path is entirely unchanged, so I am surprised it change during upgrade - very possible its a different root cause. The logs I am referring to are on pilot not the proxy.

**Comment by pmoncadaisla** [comment]
You are right @howardjohn , I was looking for logs in a cluster were logs from pilot are excluded from ingestion at Stackdriver.

I've checked for the same one I reported before, and I can see what you described:

```
2020-02-05T12:10:59.408811Z info Handling event update for pod other-deployment-6d94d5f9db-8mt4d in namespace wrong-namespace -> 10.196.0.20
2020-02-05T12:12:50.781199Z info Handling event update for pod other-deploymenti-6d94d5f9db-8mt4d in namespace wrong-namespace  -> 10.196.0.9
2020-02-05T15:24:21.009595Z info Handling event update for pod a-deployment-794f446794-clvbg in namespace correct-namespace -> 10.196.0.20
```

In this case IP address `10.196.0.20` is being reused as you described.

I guess our options if we are affected by this issue are:
- Shutdown preemtible nodes
- Wait for 1.5 and expect #20676 to be included

**Comment by howardjohn** [comment]
@pmoncadaisla this will be backported to 1.4 as well. If you want, you can run the dev builds that include this commit: https://github.com/istio/istio/wiki/Dev%20Builds. You can see all commits at https://github.com/istio/istio/commits/release-1.4. Note that we don't validate these much, so it would be at your own risk

**Comment by pmoncadaisla** [comment]
Hi @howardjohn , we have removed our preemptible nodes until the fix is included in a release.
We will be monitoring and we will keep you informed if we see the same behaviour after removing preemptible nodes.

If we don't have problems, then our issue is not related to @jasonwoodman-ascend 's

**Comment by howardjohn** [comment]
Its possible @jasonwoodman-ascend is the same issue without pre-emptible nodes. The symptom here is a running pod has its IP address changed. I've only seen it on node restarts, but perhaps it could happen in other cases?

**Comment by jasonwoodman-ascend** [comment]
@howardjohn  @pmoncadaisla I think my issue can be explained by the same root cause.  In our case, though the nodes are long living, we do restart them for patching.  I am guessing our issue started in our last patching cycle, the timing seems about right.

At this point I am waiting for the fix to be released and will re-evaluate then as I think it will most likely fix it.

**Comment by infa-rbliznet** [comment]
I think, we have same or very similar issue, so that's why I'm not filing a new bug. In our case as well istio pickups listeners from completely different namespace (kube-system) instead of namespace where service is deployed. However, it's intermittent issue, we have to run delete-create loop for deployment/service to reproduce it. The only thing that helped as is to manually define Sidecar object and repeat all the ports defined in Service object.

**Comment by howardjohn** [comment]
This should be fixed in 1.4+ now. If you still see this issue on these versions, please reopen/open another issue. Thanks!
